File: hour_score.py
"""Versioned, retrospective one-hour scoring; never controls model execution."""
import datetime as dt
import json
import time
import repair_runs
import score_weights
import scoring_policy

VERSION = 'hour-v1'
WINDOW_S = 3600


def score(job, rows, expected=(), now=None):
    now = time.time() if now is None else now
    expected = expected or job.get('expected', [])
    tids = set(job.get('tasks') or job.get('order') or [t['task'] for t in expected])
    vision = {t['task']: bool(t.get('vision')) for t in expected}
    if job.get('repair') and not job.get('_repair_rows_materialized'):rows=repair_runs.composite_rows(job,rows)
    raw = [r for r in rows if r.get('evaluation_id') == job.get('id') and r.get('task') in tids]
    intervals = job.get('active_intervals')
    timing_known = not job.get('hour_timing_unknown') and (not job.get('resume_count') or bool(intervals))
    started = job.get('started')
    if not intervals:
        intervals = [{'start': started, 'end': job.get('ended')} ] if started else []

    def active_at(ts):
        return sum(max(0, min(ts, part.get('end') or ts) - part['start']) for part in intervals)

    elapsed = active_at(job.get('ended') or now) if timing_known else job.get('progress', {}).get('elapsed_s', 0)
    elapsed += (job.get('repair') or {}).get('base_elapsed_s',0) if timing_known else 0
    within, after, errors, unknown = [], set(), 0, 0
    timeouts = set()
    for r in raw:
        if r.get('score_reason') in ('five_wrong_in_row', 'wrong_streak_limit'):
            continue
        try:
            ts = dt.datetime.fromisoformat(r['ts']).timestamp()
        except (KeyError, TypeError, ValueError):
            unknown += 1
            continue
        if not r.get('repair_inherited') and (not intervals or ts < intervals[0]['start']):
            unknown += 1
            continue
        row_elapsed=repair_runs.elapsed_at(job,r) if job.get('repair') else active_at(ts)
        if row_elapsed > WINDOW_S:
            if r.get('status') == 'completed':
                after.add(r['task'])
            continue
        if r.get('status') == 'timeout':
            timeouts.add(r['task'])
        elif r.get('status') == 'error':
            errors += 1
        elif r.get('status') == 'completed':
            within.append(r)
    available = timing_known and not unknown and not job.get("results_reset")
    correct = {r['task'] for r in within if r.get('solved')}
    completed = {r['task'] for r in within}
    net_policy = scoring_policy.is_net(job.get('scoring_policy'))
    wrong = {r['task'] for r in within if scoring_policy.incorrect(r)} - correct if net_policy else completed - correct
    abstained = {r['task'] for r in within if r.get('score_reason') == 'abstained'} - correct - wrong
    unsupported = {r['task'] for r in within if r.get('score_reason') == 'unsupported_vision'} - correct - wrong
    weights={t['task']:t.get('weight',score_weights.weight(t)) for t in expected}
    weighted=lambda ids:round(sum(weights.get(tid,1.0) for tid in ids),6) if available else None
    gross = weighted(correct)
    net = round(gross-len(wrong),6) if available else None
    lanes = {}
    for lane in ('text', 'vision'):
        rs = [r for r in within if vision.get(r['task'], r.get('kind') == 'chart-vqa' or r.get('section') in ('chart', 'charts')) == (lane == 'vision')]
        lanes[lane] = {'points': len({r['task'] for r in rs if r.get('solved')}) if available else None,
                       'weighted_points':(round(weighted({r['task'] for r in rs if r.get('solved')})-len(wrong & {r['task'] for r in rs}),6) if available else None) if net_policy else weighted({r['task'] for r in rs if r.get('solved')}),
                       'completed_questions': len({r['task'] for r in rs})}
    discovery = {}
    for domain in ('games', 'hourglass', 'dsg'):
        group = {t['task'] for t in expected if t.get('section') == 'repository_discovery' and t.get('discovery_domain') == domain} & tids
        if group:
            discovery[domain] = {'points': len(correct & group) if available else None,
                                 'weighted_points': round(weighted(correct & group) - (len(wrong & group) if net_policy else 0), 6) if available else None,
                                 'completed_questions': len(completed & group), 'total_questions': len(group)}
    repeats={t['task']:t.get('repeat',1) for t in expected}
    all_finished=bool(tids) and all(tid in timeouts or set(range(1,repeats.get(tid,1)+1)).issubset({r.get('run',1) for r in within if r['task']==tid}) for tid in tids)
    final = available and (elapsed >= WINDOW_S or all_finished)
    state = 'unavailable' if not available else 'final' if final else 'in_progress' if job.get('state') == 'running' else 'not_started' if not started else 'partial'
    return {'version': VERSION, 'window_s': WINDOW_S, 'points': len(correct) if available else None,
            'weighted_version':score_weights.VERSION,'weighted_points':net if net_policy else gross,
            'scoring_policy':job.get('scoring_policy') or scoring_policy.LEGACY,'gross_points':gross,'net_points':net if net_policy else None,
            'penalty_points':len(wrong) if net_policy else 0,'abstained_questions':len(abstained),'unsupported_questions':len(unsupported),
            'correct_tasks': sorted(correct) if available else [], 'breakdown': lanes,
            'repository_discovery': discovery,
            'timeouts':len(timeouts), 'resolved_questions':len(completed | timeouts),
            'question_timeout_policy':job.get('question_timeout_policy'),
            'completed_questions': len(completed), 'incorrect_questions': len(wrong),
            'total_questions': len(tids), 'after_deadline_questions': len(after), 'errors': errors,
            'elapsed_s': elapsed, 'remaining_s': max(0, WINDOW_S - elapsed), 'state': state,
            'repair_policy':(job.get('repair') or {}).get('policy'),
            'timing_note': 'Selected question results were reset; rerun them with a fresh clock. The original hourly score is withheld.' if job.get('results_reset') else repair_runs.NOTE if job.get('repair') and available else 'Recorded completion timestamps on active wall clock, including thinking, tools and retries.' if available else 'Missing completion timestamps or historical pause intervals; hourly score withheld.'}


def leaderboard(root, rows):
    lines = ['## Hourglass score · weighted points in one hour', '',
             'Correct questions within 3,600 active seconds earn 1–2 points on fixed section difficulty scales. Raw correct counts remain visible. '
             'Scores are not extrapolated. Compare the same bank, order and repeat policy. '
             'Metric: hour-v1; execution versions remain unchanged.', '',
             '| model | run | questions | weighted score | raw correct | text points | vision points | status |',
             '|---|---|---|---|---|---|---|---|']
    for p in sorted((root / 'evaluations').glob('*.json')):
        m = json.loads(p.read_text())
        if not m.get('id') or not m.get('expected'):
            continue
        h = score(m, rows,score_weights.enrich(root,m['expected']))
        value = lambda n: '—' if n is None else str(n)
        lines.append(f"| {m['model']} | {m['id'][:8]} | {h['total_questions']} | {value(h['weighted_points'])} | {value(h['points'])} | {value(h['breakdown']['text']['weighted_points'])} | {value(h['breakdown']['vision']['weighted_points'])} | {h['state']} |")
    return '\n'.join(lines) + '\n\n'

File: ui/hour-score.js
'use strict';
function questionWeight(t){
  if(t.section==='challenge')return 2;
  if(['math','math_logic'].includes(t.section))return {advanced_high_school:1,advanced_undergraduate:1.5,graduate:2}[t.level]??1;
  const max=['chart','charts'].includes(t.section)||t.kind==='chart-vqa'?10:t.section==='games'?5:null;
  return max&&typeof t.tier==='number'&&Number.isFinite(t.tier)?Math.round((1+(Math.min(max,Math.max(1,t.tier))-1)/(max-1))*1e6)/1e6:1;
}
// Mirrors hour_score.py; shared boundary fixtures test Python/browser parity.
function hourScore(job, rows, expected=[], now=Date.now()/1000){
  expected=expected.length?expected:(job.expected||[]);
  const tids=new Set(job.tasks||job.order||expected.map(t=>t.task));
  const vision=new Map(expected.map(t=>[t.task,!!t.vision]));
  if(job.repair&&!job._repair_rows_materialized){const own=rows.filter(r=>r.evaluation_id===job.id);rows=[...job.repair.retained.filter(r=>!own.some(x=>x.run_id===r.run_id&&x.repair_inherited)).map(r=>({...r,evaluation_id:job.id,repair_inherited:true})),...own];}
  const raw=rows.filter(r=>r.evaluation_id===job.id&&tids.has(r.task));
  let intervals=job.active_intervals;
  const timingKnown=!job.hour_timing_unknown&&(!job.resume_count||!!intervals?.length),started=job.started;
  if(!intervals?.length)intervals=started?[{start:started,end:job.ended}]:[];
  const activeAt=ts=>intervals.reduce((n,p)=>n+Math.max(0,Math.min(ts,p.end||ts)-p.start),0);
  const elapsed=timingKnown?activeAt(job.ended||now)+(job.repair?.base_elapsed_s||0):(job.progress?.elapsed_s||0);
  const events=[],within=[],after=new Set(),timeouts=new Set();let errors=0,unknown=0;
  for(const r of raw){
    if(['five_wrong_in_row','wrong_streak_limit'].includes(r.score_reason))continue;
    const ts=Date.parse(r.ts)/1000;
    if(!Number.isFinite(ts)||!r.repair_inherited&&(!intervals.length||ts<intervals[0].start)){unknown++;continue}
    const rowElapsed=r.repair_inherited?r.repair_elapsed_s:(job.repair?.base_elapsed_s||0)+activeAt(ts);
    if(rowElapsed>3600){if(r.status==='completed')after.add(r.task);continue}
    if(r.status==='timeout')timeouts.add(r.task);else if(r.status==='error')errors++;else if(r.status==='completed'){within.push(r);events.push([rowElapsed,r]);}
  }
  if(job.scoring_policy==='net-hour-v3'){const effective=new Map();for(const r of within)effective.set(r.task+'|'+(r.run??1),r);within.splice(0,within.length,...effective.values())}
  const available=timingKnown&&!unknown&&!job.results_reset,correct=new Set(within.filter(r=>r.solved).map(r=>r.task)),completed=new Set(within.map(r=>r.task));
  const netPolicy=['net-hour-v1','net-hour-v2','net-hour-v3'].includes(job.scoring_policy),neutral=['unsupported_vision','abstained','question_timeout','not_attempted','turn_limit','stopped','unfinished','wrong_streak_limit','five_wrong_in_row'];
  const wrong=new Set(within.filter(r=>!r.solved&&(!netPolicy||!neutral.includes(r.score_reason)&&!neutral.includes(r.termination))).map(r=>r.task).filter(t=>!correct.has(t)));
  const countReason=reason=>new Set(within.filter(r=>r.score_reason===reason).map(r=>r.task).filter(t=>!correct.has(t)&&!wrong.has(t))).size;
  const weights=new Map(expected.map(t=>[t.task,t.weight??questionWeight(t)]));
  const weighted=ids=>available?Math.round([...ids].reduce((n,id)=>n+(weights.get(id)??1),0)*1e6)/1e6:null;
  const lanes={};
  for(const lane of ['text','vision']){
    const rs=within.filter(r=>(vision.has(r.task)?vision.get(r.task):r.kind==='chart-vqa'||['chart','charts'].includes(r.section))===(lane==='vision'));
    const correctLane=new Set(rs.filter(r=>r.solved).map(r=>r.task));
    lanes[lane]={points:available?correctLane.size:null,weighted_points:available?Math.round((weighted(correctLane)-(netPolicy?[...new Set(rs.map(r=>r.task))].filter(t=>wrong.has(t)).length:0))*1e6)/1e6:null,completed_questions:new Set(rs.map(r=>r.task)).size};
  }
  const repeats=new Map(expected.map(t=>[t.task,t.repeat??1]));
  const discovery={};
  for(const domain of ['games','hourglass','dsg']){
    const group=expected.filter(t=>t.section==='repository_discovery'&&t.discovery_domain===domain&&tids.has(t.task)).map(t=>t.task);
    if(group.length)discovery[domain]={points:available?group.filter(t=>correct.has(t)).length:null,weighted_points:available?Math.round((weighted(new Set(group.filter(t=>correct.has(t))))-(netPolicy?group.filter(t=>wrong.has(t)).length:0))*1e6)/1e6:null,completed_questions:group.filter(t=>completed.has(t)).length,total_questions:group.length};
  }
  const cumulative=job.scoring_policy==='net-hour-v3';
  const attemptMap=new Map();for(const r of within)attemptMap.set(r.task+'|'+(r.run??1),r);
  const attempts=[...attemptMap.values()],correctAttempts=attempts.filter(r=>r.solved),wrongAttempts=attempts.filter(r=>!r.solved&&!neutral.includes(r.score_reason)&&!neutral.includes(r.termination));
  const gross=cumulative?(available?correctAttempts.reduce((n,r)=>n+(weights.get(r.task)??1),0):null):weighted(correct);
  const penalties=netPolicy?(cumulative?wrongAttempts.length:wrong.size):0;
  const total=available?Math.round((gross-penalties)*1e6)/1e6:null;
  if(cumulative){
    const apply=(sub,rs)=>{sub.points=available?rs.filter(r=>r.solved).length:null;sub.weighted_points=available?Math.round(rs.reduce((n,r)=>n+(r.solved?(weights.get(r.task)??1):!neutral.includes(r.score_reason)&&!neutral.includes(r.termination)?-1:0),0)*1e6)/1e6:null;};
    for(const lane of ['text','vision'])apply(lanes[lane],attempts.filter(r=>(vision.has(r.task)?vision.get(r.task):r.kind==='chart-vqa'||['chart','charts'].includes(r.section))===(lane==='vision')));
    for(const [domain,sub] of Object.entries(discovery)){const ids=new Set(expected.filter(t=>t.discovery_domain===domain).map(t=>t.task));apply(sub,attempts.filter(r=>ids.has(r.task)));}
  }
  const allFinished=tids.size>0&&[...tids].every(tid=>timeouts.has(tid)||Array.from({length:repeats.get(tid)??1},(_,i)=>i+1).every(n=>within.some(r=>r.task===tid&&(r.run??1)===n)));
  const final=available&&(elapsed>=3600||(allFinished&&!job.round_policy));
  const state=!available?'unavailable':final?'final':job.state==='running'?'in_progress':!started?'not_started':'partial';
  const horizon=final?3600:Math.max(0,Math.min(3600,elapsed));
  const totalAvailable=Math.round([...tids].reduce((n,tid)=>n+(weights.get(tid)??1),0)*1e6)/1e6;
  return {version:'hour-v1',window_s:3600,points:available?(cumulative?correctAttempts.length:correct.size):null,correct_tasks:available?[...correct].sort():[],breakdown:lanes,repository_discovery:discovery,
    score_version:'total-points-v1',hourglass_score:total,total_available_points:totalAvailable,
    weighted_version:'weighted-hour-v1',weighted_points:total,
    scoring_policy:job.scoring_policy??'weighted-hour-v1',gross_points:gross,net_points:netPolicy?total:null,penalty_points:penalties,abstained_questions:countReason('abstained'),unsupported_questions:countReason('unsupported_vision'),
    timeouts:timeouts.size,resolved_questions:new Set([...completed,...timeouts]).size,resolved_available_points:weighted(new Set([...completed,...timeouts])),question_timeout_policy:job.question_timeout_policy??null,
    completed_questions:completed.size,incorrect_questions:cumulative?wrongAttempts.length:wrong.size,total_questions:tids.size,
    after_deadline_questions:after.size,errors,elapsed_s:elapsed,remaining_s:Math.max(0,3600-elapsed),state,
    repair_policy:job.repair?.policy??null,
    timing_note:job.results_reset?'Selected question results were reset; rerun them with a fresh clock. The original hourly score is withheld.':job.repair&&available?job.repair.note:available?'Recorded completion timestamps on active wall clock, including thinking, tools and retries.':'Missing completion timestamps or historical pause intervals; hourly score withheld.'};
}
if(typeof module!=='undefined')module.exports={hourScore,questionWeight};
